### ISSUE TITLE
[Refactor] `Trainer` class does not support custom `optimize_step()` with additional arguments

### ISSUE THREAD

**Issue by DanieleCucurachi** [pasqal-io/qadence#732]
## Issue Description

The `optimize_step()` function in the `Trainer` class controls low-level training operations and enables advanced loss functions and loss balancing techniques. However, the current implementation does not allow users to pass arbitrary arguments to the `optimize_step()` function.

#### Current Code:
```python
class Trainer(BaseTrainer):

...

@BaseTrainer.callback("train_batch")
def run_train_batch(
    self, batch: tuple[torch.Tensor, ...]
) -> tuple[torch.Tensor, dict[str, Any]]:
    """
    Runs a single training batch, performing optimization.

    Depending on `use_grad`, the optimization method is chosen:
    - `use_grad=True`: Uses gradient-based optimization with `optimize_step`.
    - `use_grad=False`: Uses gradient-free optimization with `update_ng_parameters`.

    Args:
        batch (tuple[torch.Tensor, ...]): A batch of data from the DataLoader.

    Returns:
        tuple[torch.Tensor, dict[str, Any]]: Loss and metrics for the batch.
    """
    if self.use_grad:
        # Perform gradient-based optimization
        loss_metrics = self.optimize_step(
            model=self.model,
            optimizer=self.optimizer,
            loss_fn=self.loss_fn,
            xs=batch,
            device=self.accelerator.execution.device,
            dtype=self.accelerator.execution.data_dtype,
        )
   
    ...

```

Currently, if users want to implement a custom optimize_step() function with additional arguments, they must use workarounds like this:

```python
...

additional_argument_1 = ...
additional_argument_2 = ...

...

def wrapper_optimize_step(model, optimizer, loss_fn, xs, device, dtype):
    return custom_optimize_step(
        model, optimizer, loss_fn, device, dtype, additional_argument_1, additional_argument_2, ...
    )

...

trainer = Trainer(model, optimizer, ... , optimize_step=wrapper_optimize_step, ... )
```

## Desired Behavior:

It would be much more user friendly if the `Trainer` class allowed users to provide additional, unspecified arguments directly when calling the `optimize_step()` function. This would eliminate the need for workarounds and offer more flexibility.

```python
...

additional_argument_1 = ...
additional_argument_2 = ...

...

trainer = Trainer(model, optimizer, ... , optimize_step=custom_optimize_step, add_args={"additional_argument_1": additional_argument_1, ...}, ... )
```

**NB:** This is just an example. The final implementation doesn't have to follow this exactly, using args or kwargs to pass additional arguments might be a more flexible approach.

**Comment by mlahariya** [comment]
@DanieleCucurachi What's the priority on this issue?

**Comment by DanieleCucurachi** [comment]
@mlahariya if we could close this before the end of the next week it would be great.
